File: 29-divide-two-integers/divide-two-integers.c
int divide(int dividend, int divisor) {
    // Handle the case where dividend is equal to divisor 
    if (dividend == divisor) return 1;

    unsigned int ans = 0;
    int sign = 1; 

    //Determine the sign of the result 
    if ((dividend < 0 && divisor > 0) || (dividend > 0 && divisor < 0 ))
        sign = -1; 
    
    //Convert both dividend and divisor to positive integers 
    //unsigned int n = abs(dividend) , d = abs(divisor);
    unsigned int n = (dividend < 0) ? (unsigned int)(-(long long)dividend) : (unsigned int)dividend;
    unsigned int d = (divisor < 0) ? (unsigned int)(-(long long)divisor) : (unsigned int)divisor;


    //Perform the actual divison 
    while (n >=d){
        int count = 0; 
        while (n > (d << (count+1))){
            count++;
        }
        n -= d << count; 
        ans += 1 << count; 
    }

    //Handle overflow case 
    if (ans == (1U << 31) && sign == 1) return INT_MAX; 


    return sign * ans; 
}

File: 29-divide-two-integers/divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        
        # Handle overflow
        if dividend == INT_MIN and divisor == -1:
            return INT_MAX
        
        # Determine sign
        negative = (dividend < 0) ^ (divisor < 0)
        
        # Work with positive numbers
        dividend, divisor = abs(dividend), abs(divisor)
        result = 0
        
        while dividend >= divisor:
            temp, multiple = divisor, 1
            while dividend >= (temp << 1):
                temp <<= 1
                multiple <<= 1
            dividend -= temp
            result += multiple
        
        # Apply sign
        result = -result if negative else result
        
        # Clamp to 32-bit range
        return max(INT_MIN, min(INT_MAX, result))